dds/DCPS/transport/rtps_udp/RtpsUdpReceiveStrategy.cpp

```diff
@@ -352,6 +352,13 @@ RtpsUdpReceiveStrategy::deliver_sample_i(ReceivedDataSample& sample,
     link_->received(data, receiver_.source_guid_prefix_);
     recvd_sample_ = 0;
 
+    const RepoIdSet* readers = link_->updateWriterSeqReaders(sample.header_.publication_id_, sample.header_.sequence_);
```

Comment:
This will crash if there's an association occuring at the same time, since both threads will be accessing / changing the internal RepoIdSet at the same time. I'd change the method signature of updateWriterSeqReaders (or whatever we call it) to take in readers_withheld_ and then insert the GUIDs from the internal set while the RtpsUdpDataLink lock is held.